src/workerd/jsg/jsg.c++

```diff
@@ -55,6 +55,20 @@ void Data::destroy() {
   if (isolate != nullptr) {
     if (v8::Locker::IsLocked(isolate)) {
       handle.Reset();
+
+      // If we have a TracedReference, Reset() it too, to let V8 know that this value is no longer
+      // used. Note that merely detroying the TracedReference does nothing -- only explicitly
+      // calling Reset() has an effect.
+      //
+      // In particular, this permits `Data` values to be collected by minor (non-tracing) GC, as
+      // long as there are no cycles.
+      //
+      // TODO(now): Determine if it's always safe to reset the handle here. The handle becomes
```

Comment:
I'm pretty sure this is not safe, particularly what I discovered in the next commit: that cppgc objects are actually destroyed some time after the tracing GC has supposedly finished.

What I can do here instead is use some sort of thread-local to determine if we were triggered by a shim destructor vs. a minor GC. In the case of a shim destructor, we can't reset the handle. In the case of a minor GC, we can.

What about when the case where neither is happening? That is, the object isn't being collected by GC, but rather is being destroyed by application code, e.g. because it's discarding the handle. I think in that case, it is safe to reset the reference, because it's clearly still reachable.

So: Set a thread-local flag during CppShim destructor, and don't reset the reference in that case.